ggerganov/whisper.cpp
# Testing : Compare CPU backend with GPU backend

**Phenomenon:** When utilizing the CUDA backend, the transcription tends to produce hallucinations. See #1688 for more details.

**Audio for error reproduction:** [01-03(Easy to Learn Chinese + Second Edition + Textbook 2).zip](https://github.com/ggerganov/whisper.cpp/files/13780212/01-03.Easy.to.Learn.Chinese.%2B.Second.Edition.%2B.Textbook.2.zip)

**How to Reproduce the Error:**

1. First, grab the latest code from the master branch.
2. Use a computer that's equipped with an NVIDIA graphics card.
3. Compile the code with `CMake`, ensuring to include the `-DWHISPER_CUBLAS=1` option in the process.
4. Execute the main program, using the sample audio I've provided, the `whisper-large-v2` model, and the `-l zh` flag.
5. At this point, you'll notice some hallucinations in the output.
6. Now, try running main again, just like in `step 4`, but this time add the `-ng` flag. You'll see that the hallucinations no longer occur.

This PR introduces code that compares the outputs of different backends with the CPU backend for each tensor operation. It executes the encoder once and does not produce any transcriptions.
